Utility/DynamicJsonParser.cs

```diff
@@ -0,0 +1,237 @@
+﻿//
+//  Copyright (c) Microsoft Corporation. All rights reserved.
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+namespace Microsoft.PackageManagement.NuGetProvider
+{
+    using System;
+    using System.Collections;
+    using System.Collections.Generic;
+    using System.Dynamic;
+    using System.Linq;
+    using System.Management.Automation;
+    using System.Reflection;
+
+
+    /// <summary>
+    /// Using PowerShell instance, parse valid JSON into a dynamic object.
+    /// </summary>
+    public class DynamicJsonParser
+    {
+        private static object psCreateLock = new object();
+        private static System.Management.Automation.PowerShell powershell = null;
+
+        private static System.Management.Automation.PowerShell PowerShellInstance
+        {
+            get
+            {
+                if (powershell == null)
+                {
+                    lock (psCreateLock)
+                    {
+                        if (powershell == null)
+                        {
+                            powershell = Init();
+                        }
+                    }
+                }
+
+                return powershell;
+            }
+        }
+
+        private static System.Management.Automation.PowerShell Init()
+        {
+            return System.Management.Automation.PowerShell.Create();
+        }
+
+        private DynamicJsonParser() { }
+
+        static DynamicJsonParser() { }
+
+        /// <summary>
+        /// Parse valid JSON into a dynamic object. All properties starting with @ or $ are added to the Metadata property object.
+        /// </summary>
+        /// <param name="json">JSON to parse.</param>
+        /// <param name="numericalPrefix">Prefix for properties that start with a number.</param>
+        /// <param name="lowercaseProperties">Lowercase all properties when you don't know the casing of the input JSON.</param>
+        /// <returns>Dynamic object.</returns>
+        public static dynamic Parse(string json, string numericalPrefix = "n", bool lowercaseProperties = true)
```

Comment:
dynamic will be much harder to diagnosis for runtime if any issues occurs.

Reply:
Not sure what the justification is for this opinion. If anything, parse errors will be easier to debug vs. strongly-typed object model. Currently translating from PSObject is cleaner into a dynamic rather than a strong type, and "parse" errors show as "missing property 'aProperty'" rathern than something like "failed to deserialize into class X"

---
In reply to: [154462234](https://github.com/OneGet/NuGetProvider/pull/25#discussion_r154462234) [](ancestors = 154462234)